src/grids.jl

```diff
@@ -466,6 +565,126 @@ get_colat(G::Type{<:HEALPixGrid},nside::Integer) =
 full_grid(::Type{<:HEALPixGrid}) = FullHEALPixGrid    # the full grid with same latitudes
 matrix_size(G::HEALPixGrid) = (5G.nside,5G.nside)
 
+"""
+    J = HEALPix4Grid{T}
+
+A HEALPix4 grid with 4 base faces, each `nside`x`nside` grid points, each covering the same area.
+The values of all grid points are stored in a vector field `v` that unravels the data 0 to 360˚,
+then ring by ring, which are sorted north to south."""
+struct HEALPix4Grid{T} <: AbstractHEALPix4Grid{T}
+    data::Vector{T}    # data vector, ring by ring, north to south
+    nside::Int      # nside^2 is the number of pixel in each of the 12 base pixel
+
+    HEALPix4Grid{T}(data,nside) where T = length(data) == npoints_healpix4(nside) ?
+    new(data,nside) : error("$(length(data))-element Vector{$(eltype(data))}"*
+    "cannot be used to create an J$nside HEALPix4Grid{$T}.")
+end
+
+npoints_healpix4(nside::Integer) = 4nside^2
+nside_healpix4(npoints::Integer) = round(Int,sqrt(npoints/4))  # inverse of npoints_healpix4
+nlat_healpix4(nside::Integer) = 2nside-1
+nlon_healpix4(nside::Integer,j::Integer) = 4j
+
+# infer nside from data vector length, infer parametric type from eltype of data
+HEALPix4Grid{T}(data::AbstractVector) where T = HEALPix4Grid{T}(data,nside_healpix4(length(data)))
+HEALPix4Grid(data::AbstractVector,n::Integer...) = HEALPix4Grid{eltype(data)}(data,n...)
+
+truncation_order(::Type{<:HEALPix4Grid}) = 3                 # cubic (in longitude)
+get_truncation(::Type{<:HEALPix4Grid},nside::Integer) = nside-1
+get_resolution(::Type{<:HEALPix4Grid},trunc::Integer) = roundup_fft(trunc+1)
+get_colat(G::Type{<:HEALPix4Grid},nside::Integer) =
+    ( colat_half=[acos(1-(j/nside)^2) for j in 1:nside]; vcat(colat_half,π.-colat_half[end-1:-1:1]) )
+# get_lon() is already implemented for AbstractHEALPix4Grid
+# get_colatlons() is already implemented for AbstractHEALPix4Grid
+# each_index_in_ring() is already implemented for AbstractHEALPix4Grid
+full_grid(::Type{<:HEALPix4Grid}) = FullHEALPix4Grid    # the full grid with same latitudes
+
+matrix_size(G::HEALPix4Grid) = (2G.nside,2G.nside)
+matrix_size(::Type{HEALPix4Grid},nside::Integer)=(2nside,2nside)
+Base.Matrix(G::HEALPix4Grid{T};kwargs...) where T = Matrix!(zeros(T,matrix_size(G)...),G;kwargs...)
+"""
+    Matrix!(M::AbstractMatrix,
+            G::HEALPix4Grid;
+            quadrant_rotation=(0,1,2,3),
+            matrix_quadrant=((2,2),(1,2),(1,1),(2,1)),
+            )
+
+Sorts the gridpoints in `G` into the matrix `M` without interpolation.
+Every quadrant of the grid `G` is rotated as specified in `quadrant_rotation`,
+0 is no rotation, 1 is 90˚ clockwise, 2 is 180˚ etc. Grid quadrants are counted
+eastward starting from 0˚E. The grid quadrants are moved into the matrix quadrant
+(i,j) as specified. Defaults are equivalent to centered at 0˚E and a rotation
+such that the North Pole is at M's midpoint."""
+Matrix!(M::AbstractMatrix,G::HEALPix4Grid;kwargs...) = Matrix!((M,G);kwargs...)
```

Comment:
We can probably define this for any `AbstractGrid`?

Reply:
If we can, that would be great. I don't see how we can do this for HEALPix12. Any ideas?